src/index.js

```diff
@@ -364,114 +364,147 @@ class PayfitContentScript extends ContentScript {
   async fetchPayslips({ context, fetchedDatesArray, i }) {
     this.log('info', '📍️ fetchPayslips starts')
     await this.navigateToPayrollsPage()
-    const numberOfBills = await this.evaluateInWorker(
-      function getNumberOfBills() {
+    await this.runInWorkerUntilTrue({
+      method: 'getPayslipsInfos'
+    })
+    const alreadyFetchedIds = []
+    const allPayslipsIds = this.store.contractBillsInfos.payslipsIds
+    const limit = 10
+    const totalIdsLength = allPayslipsIds.length
+    this.log('info', `totalIdsLength : ${totalIdsLength}`)
+    for (let j = totalIdsLength; j > 0; j -= limit) {
+      const group = Array.from(allPayslipsIds).slice(Math.max(j - limit, 0), j)
+      const fetchedIds = await this.runInWorker('getBatchOfTen', {
+        limit,
+        group
+      })
+      alreadyFetchedIds.push(...fetchedIds)
+      await this.runInWorkerUntilTrue({
+        method: 'checkInterception',
+        args: [{ type: 'bills', number: fetchedIds.length }]
+      })
+      const billsBatch = await this.runInWorker('getBills')
+      let subPath = await this.determineSubPath(fetchedDatesArray, i)
+      // RECOLLER LE BLOC SAVEBILLS
+      await this.saveFiles(billsBatch, {
+        context,
+        fileIdAttributes: ['vendorId'],
+        contentType: 'application/pdf',
+        qualificationLabel: 'pay_sheet',
+        subPath
+      })
+    }
+    await this.runInWorker('emptyInterceptionsArrays')
+  }
+
+  async getBatchOfTen(options) {
+    this.log('info', '📍️ getBatchOfTen starts')
+    const payslipsIds = []
+    await waitFor(
+      () => {
+        const foundElements = document.querySelectorAll(
+          'div[data-testid*="payslip-"] > div'
+        )
+        const foundIds = Array.from(foundElements).map(element =>
+          element.parentNode.getAttribute('data-testid')
+        )
+        if (foundIds.some(entry => entry.includes(options.group[0]))) {
+          this.log('info', 'found first id in html')
+          return true
+        } else {
+          this.log('info', 'found nothing, scrolling')
+          const beforeLast = document.querySelector(
+            '.ReactVirtualized__Grid__innerScrollContainer > div:nth-last-child(5)'
+          )
+          beforeLast.scrollIntoView({ behavior: 'instant', block: 'end' })
+          return false
+        }
+      },
+      {
+        interval: 1000,
+        timeout: 30 * 1000
+      }
+    )
+    const sectionBillsElements = document.querySelectorAll(
+      'div[data-testid*="payslip-"] > div'
+    )
+    const neededPayslips = []
+    for (const billElement of sectionBillsElements) {
+      const elementId = billElement.parentNode.getAttribute('data-testid')
+      if (options.group.includes(elementId)) {
+        neededPayslips.push(billElement)
+      }
+    }
+    for (let i = 0; i < neededPayslips.length; i++) {
```

Comment:
There are a lot of loops and it is really hard to follow. I think at least each loop should be replaced by a function with a proper name. It may make it easier to follow.